### PR TITLE
First Draft of Lecture 10

### PR DESCRIPTION
**Pull request by sid-ramakrishnan**
Hi Prof Mueller!

I have added the content for lecture 10 on remark. Please let me know where I can make any additional changes if any.

Thanks

### CODE REVIEW

**Comment by amueller** [comment]
Thanks Siddharth. Generally I prefer if the images have semantic names, though you can leave it as-is for this one.

File: slides/aml-10-021918-trees-forests/index.html [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+ - Minimize “impurity”
+]
+.column[
+![:scale 50%](images/img_3.png)
+---
+![:scale 50%](images/img_4.png)
+]
+
+???
+Does the layout look okay??
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html:168]: yeah probably fine. A bit crowded. I might change this a bit later.

File: slides/aml-10-021918-trees-forests/index.html [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+![:scale 70%](images/img_2.png)
+ - Continuous features:
+ - “questions” are thresholds on single features.
+ - [Other methods are possible but not as common]
+ - For each split: exhaustive search over all features and thresholds!
+ - Minimize “impurity”
+]
+.column[
+![:scale 50%](images/img_3.png)
+---
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: I don't understand what you're doing here ``---`` is new slide, right?

File: slides/aml-10-021918-trees-forests/index.html [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+???
+
+---
+
+# Visualizing trees with sklearn
+
+.smaller[
+```python
+from sklearn.datasets import load_breast_cancer
+cancer = load_breast_cancer()
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html:224]: If there is no output in between you can make it one big code block.

File: slides/aml-10-021918-trees-forests/index.html [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+
+???
+---
+
+.smaller[
+```python
+from sklearn.model_selection import GridSearchCV
+param_grid = {'max_depth':range(1, 7)}
+grid = GridSearchCV(DecisionTreeClassifier(random_state=0),param_grid=param_grid,
+cv=10)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: please use pep8 indentation. In this case, line up with DecisionTree.

File: slides/aml-10-021918-trees-forests/index.html [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+]
+
+???
+---
+.smaller[
+```python
+from sklearn.model_selection import GridSearchCV
+param_grid = {'max_leaf_nodes':range(2, 20)}
+grid = GridSearchCV(DecisionTreeClassifier(random_state=0),
+param_grid=param_grid, cv=10)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: indentation...

File: slides/aml-10-021918-trees-forests/index.html [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+???
+---
+.center[
+![:scale 95%](images/img_20.png)
+]
+???
+---
+#Extrapolation
+.center[
+![:scale 85%](images/img_20.png)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: this is missing the predictions.

File: slides/aml-10-021918-trees-forests/index.html [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+]
+???
+---
+
+class: split-40
+# Instability
+
+.column[
+.smaller[
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: maybe see if you can make it even smaller (I had a "smallest" class somewhere)? right now it's impossible to read :-/

File: slides/aml-10-021918-trees-forests/index.html [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+```python
+voting = VotingClassifier([('logreg', LogisticRegression(C=100)),
+('tree', DecisionTreeClassifier(max_depth=3, random_state=0))],
+voting='soft')
+voting.fit(X_train, y_train)
+lr, tree = voting.estimators_
+print(("{:.2f} " * 3).format(voting.score(X_test, y_test),
+lr.score(X_test, y_test), tree.score(X_test, y_test)))
+```
+0.88 0.84 0.80
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: output should have triple backticks around it.

File: slides/aml-10-021918-trees-forests/index.html [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+
+.smaller[
+```python
+train_scores = []
+test_scores = []
+oob_scores = []
+
+feature_range = range(1, 64, 5)
+for max_features in feature_range:
+    rf = RandomForestClassifier(max_features=max_features, oob_score=True, n_estimators=200, random_state=0)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: needs linebreak.

File: slides/aml-10-021918-trees-forests/index.html [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+    train_scores.append(rf.score(X_train, y_train))
+    test_scores.append(rf.score(X_test, y_test))
+    oob_scores.append(rf.oob_score_)
+```
+]
+
+???
+---
+.center[
+![:scale 70%](images/img_31.png)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: could be larger

File: slides/aml-10-021918-trees-forests/index.html [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+# Variable Importance
+
+.smaller[
+```python
+X_train, X_test, y_train, y_test = train_test_split(
+    iris.data, iris.target, stratify=iris.target, random_state=1)
+rf = RandomForestClassifier(n_estimators=100).fit(X_train, y_train)
+```
+```python
+rf.feature_importances_
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html]: missing output

File: slides/aml-10-021918-trees-forests/index.html [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+---
+
+class: centre,middle
+# Decision Trees for Classification
+???
+---
+
+# Idea: series of binary questions
+.center[
+![:scale 70%](images/img_1.png)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html:145]: something is wrong with this picture?

**Comment by sid-ramakrishnan** [comment]
Okay, I will make sure that I provide the names for Lecture 11

**Review by amueller — COMMENTED** [review]
Looking good, some minor nitpicks!

File: slides/aml-10-021918-trees-forests/index.html (hunk 2 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 3 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html:224]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 4 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 5 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 8 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 12 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html:145]: According to 2017 content, this was the picture provided so I used the same one. Should I add something else?

File: slides/aml-10-021918-trees-forests/index.html (hunk 7 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: I didnt find a smallest class so I added another class called tiny code which has made this smaller. Hope it looks better now. Will create a new pull request with all these changes in a little while for you to review.

File: slides/aml-10-021918-trees-forests/index.html (hunk 9 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 10 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved.

File: slides/aml-10-021918-trees-forests/index.html (hunk 11 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

File: slides/aml-10-021918-trees-forests/index.html (hunk 6 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html]: Resolved

**Comment by sid-ramakrishnan** [comment]
I have updated the slides with the latest changes. I'm not sure what img_1.png should be because the old slides have the same picture. Hope the rest of the changes are fine.

File: slides/aml-10-021918-trees-forests/index.html (hunk 12 above)
Comment (amueller) [reply slides/aml-10-021918-trees-forests/index.html:145]: It's actually this as you can see in the pdf
[02-supervised-learning_142_0.pdf](https://github.com/amueller/COMS4995-s18/files/1720681/02-supervised-learning_142_0.pdf)

Comment (amueller) [reply slides/aml-10-021918-trees-forests/index.html:145]: here's a high resolution png:
animal-tree

**Comment by amueller** [comment]
What program did you open the old slides with?

File: slides/aml-10-021918-trees-forests/index.html (hunk 12 above)
Comment (sid-ramakrishnan) [reply slides/aml-10-021918-trees-forests/index.html:145]: Thanks. I will add this right away

**Comment by sid-ramakrishnan** [comment]
I used Powerpoint to open to .odp. It just showed a big cross so I thought that was strange. In any case, I'll make sure this new one gets added and I'll push the changes along with my changes for Lecture 11

**Comment by amueller** [comment]
Powerpoint probably doesn't do a good job in rendering the odp. If something looks strange, maybe double check the pdf.

**Comment by sid-ramakrishnan** [comment]
I have incorporated the image in Lecture 10 and made the slides for 11. Its still a little rough and I have some doubts regarding how I should semantically caption some of the images? I have mentioned it in the notes section of the markdown. Could you review the slides and help me out with these and I'll polish them by Wednesday

File: slides/aml-11-022118-gradient-boosting-calibration/index.html [hunk 13]
```diff
@@ -0,0 +1,646 @@
...
+first_stage = make_pipeline(voting, reshaper)
+transform_cv = cross_val_predict(first_stage, X_train, y_train, cv=10, method="transform")
+second_stage = LogisticRegression(C=100).fit(transform_cv, y_train)
+print(second_stage.coef_)
+[[ 2.09  -1.424  7.93 ]]
+```
+
+```python
+second_stage.score(transform_cv, y_train)
+0.95999999999999996```
```
Comment (amueller) [inline slides/aml-11-022118-gradient-boosting-calibration/index.html:393]: can you please round scores to 3 decimals?

File: slides/aml-11-022118-gradient-boosting-calibration/index.html [hunk 14]
```diff
@@ -0,0 +1,646 @@
...
+fig, axs = plot_partial_dependence(gbrt, X_train, [np.argsort(gbrt.feature_importances_)[-2:]],
+                                   feature_names=boston.feature_names,
+                                   n_jobs=3, grid_resolution=50)
+```
+]
+.center[
+![:scale 70%](images/img_11.png)
+]
+.center[
+Insert Caption
```
Comment (amueller) [inline slides/aml-11-022118-gradient-boosting-calibration/index.html:220]: It's not so much about having a caption. I like to keep the slides clean and provide either video or presenter notes. The idea was to have filenames that are not ``img_11.png`` but rather ``partial_dependence_plot.png``.

Comment (sid-ramakrishnan) [reply slides/aml-11-022118-gradient-boosting-calibration/index.html:220]: Oh I see what you mean. I'll change that

**Comment by amueller** [comment]
all done, at least with lecture 10? I need to finish it tomorrow as I'm out over the weekend.

**Comment by sid-ramakrishnan** [comment]
Yes. Done with 10

**Comment by amueller** [comment]
cool, I'll merge. You can send any changes to 11 as separate PR.

**Comment by amueller** [comment]
slide 7 (tree prediction) is missing several parts of the images / animation.

File: slides/aml-10-021918-trees-forests/index.html [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+      .invisible {
+          visibility: hidden
+      }
+      .remark-code, .remark-inline-code { font-family: 'Ubuntu Mono'; }
+      .column:first-of-type {float:left}
+      .column:last-of-type {float:right}
+      .remark-code, .remark-inline-code .tiny-code{
+        font-size: 15px;
+      }
+      .split-40 .column:first-of-type {width: 50%}
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html:108]: this doesn't seem to work. The "last-of-type" is not applied. I tried to change it to 45% and it does nothing...

File: slides/aml-10-021918-trees-forests/index.html [hunk 16]
```diff
@@ -0,0 +1,724 @@
...
+X_train, X_test, y_train, y_test = 
+train_test_split(iris.data, iris.target, stratify=iris.target, random
+                _state=1)
+tree = DecisionTreeClassifier(max_leaf_nodes=6).fit(X_train,y_train)
+tree_dot = export_graphviz(tree,out_file=None,feature_names=iris.
+                          feature_names)
+graphviz.Source(tree_dot)
+```
+]
+.center[![:scale 35%](images/img_22.png)
```
Comment (amueller) [inline slides/aml-10-021918-trees-forests/index.html:391]: you can see that the last of type column didn't work here: the scale of the image is much different from the scale of the other image because the column is not restricted. I'm not sure why you didn't use left-column and right-column.
